apps/test-app/frontend/src/components/tree-widget/stateless/CustomHooks.ts

```diff
@@ -0,0 +1,76 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { useCallback, useEffect, useRef, useState } from "react";
+import { RenderedItemsRange, TreeDataProvider, TreeModelSource, TreeNodeLoader } from "@itwin/components-react";
+import { HierarchyProvider, IPrimitiveValueFormatter } from "@itwin/presentation-hierarchy-builder";
+import { startTreeReload } from "./TreeReloader";
+
+interface ControlledTreeComponents {
+  nodeLoader: TreeNodeLoader<TreeDataProvider>;
+  modelSource: TreeModelSource;
+}
+
+/** @internal */
+export function useControlledTreeComponentsState(dataProvider: TreeDataProvider) {
+  function createNewComponents(newDataProvider: TreeDataProvider): ControlledTreeComponents {
```

Comment:
Can you move `createNewComponents` outside `useControlledTreeComponentsState` so it would have stable reference? In this case it works but this approach is error prone.

Reply:
Moved outside `useControlledTreeComponentsState`